File: hard/next_number.rb
def dsig(a)
  d = 0
  while a > 0
    r = a % 10
    d += 1 << (3 * r) if r > 0
    a /= 10
  end
  d
end

File.open(ARGV[0]).each_line do |line|
  a = line.to_i
  b = a + 9
  d = dsig(a)
  b += 9 while d != dsig(b)
  puts b
end

File: hard/next_number.pl
use strict;
use integer;

sub dsig {
    my ( $a, $d ) = ( $_[0], 0 );
    while ( $a > 0 ) {
        my $r = $a % 10;
        if ( $r > 0 ) {
            $d += 1 << ( 3 * $r );
        }
        $a /= 10;
    }
    return ($d);
}

open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    next if ( $line =~ m/^\s$/ );
    my $b = $line + 9;
    my $s = dsig($line);
    while ( $s != dsig($b) ) {
        $b += 9;
    }
    printf "$b\n";
}
close(INFILE);
